src/main/java/com/minecolonies/coremod/generation/ItemNbtCalculator.java

```diff
@@ -0,0 +1,219 @@
+                list.addAll(keyObjectList);
+                keyMapping.put(resourceLocation.toString(), list);
+            }
+            else
+            {
+                keyMapping.put(resourceLocation.toString(), keyObjectList);
+            }
+        }
+
+        final Path path = packOutput.createPathProvider(PackOutput.Target.DATA_PACK, "compatibility").file(new ResourceLocation(MOD_ID, "itemnbtmatching"), "json");
+        final JsonArray jsonArray = new JsonArray();
+        for (final Map.Entry<String, Set<CheckedNbtKey>> entry : keyMapping.entrySet())
+        {
+            final JsonObject jsonObject = new JsonObject();
+            jsonObject.addProperty("item", entry.getKey());
+
+            if (!entry.getValue().isEmpty())
+            {
+                final JsonArray subArray = new JsonArray();
+                entry.getValue().forEach(key -> subArray.add(serializeKeyToJson(key)));
+                jsonObject.add("checkednbtkeys", subArray);
+            }
+
+            jsonArray.add(jsonObject);
+        }
+
+        return DataProvider.saveStable(cache, jsonArray, path);
+    }
+
+    /**
+     * Serialize a checked nbt key to json.
+     * @param keyObject the key object to serialize.
+     * @return the output json.
+     */
+    public static JsonObject serializeKeyToJson(final CheckedNbtKey keyObject)
+    {
+        final JsonObject obj = new JsonObject();
+        obj.addProperty("key", keyObject.key);
+
+        if (!keyObject.children.isEmpty())
+        {
+            final JsonArray jsonArray = new JsonArray();
+            keyObject.children.forEach(child -> jsonArray.add(serializeKeyToJson(child)));
+            obj.add("children", jsonArray);
+        }
+        return obj;
+    }
+
+    /**
+     * Create a checked nbt key from nbt.
+     * @param key the key to retrieve.
+     * @param tag the tag to deserialize it from.
+     * @return a new checked nbt key.
+     */
+    public static CheckedNbtKey createKeyFromNbt(final String key, final CompoundTag tag)
+    {
+        if (tag.get(key) instanceof CompoundTag)
+        {
+            final CompoundTag subTag = tag.getCompound(key);
+            return new CheckedNbtKey(key, subTag.getAllKeys().stream().map(subKey -> createKeyFromNbt(subKey, subTag)).collect(Collectors.toSet()));
+        }
+        else
+        {
+            return new CheckedNbtKey(key, Collections.emptySet());
+        }
+    }
+
+    /**
+     * Create a checked nbt key from json.
+     * @param jsonObject the object to serialize it from.
+     * @return the output key.
+     */
+    public static CheckedNbtKey deserializeKeyFromJson(final JsonObject jsonObject)
+    {
+        final String key = jsonObject.get("key").getAsString();
+        if (jsonObject.has("children"))
+        {
+            final Set<CheckedNbtKey> children = new HashSet<>();
+            jsonObject.getAsJsonArray("children").forEach(child -> deserializeKeyFromJson(child.getAsJsonObject()));

```

Comment:
Is there missing a `children.add` in the forEach function here? It looks like you are reading the children array out of the json, parsing each element as CheckedNBTKey, then discarding them

Reply:
Good find